opnieuw/retries.py

```diff
@@ -285,46 +249,24 @@ def decorator(
     ) -> Callable[P, Coroutine[None, None, R]]:
         @functools.wraps(f)
         async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
-            last_exception = None
-
-            retry_state = _get_retry_state_class(namespace)(
+            backoff_calculator = _get_backoff_calculator_class(namespace)(
                 MonotonicClock(),
-                max_calls_total=max_calls_total,
-                retry_window_after_first_call_in_seconds=retry_window_after_first_call_in_seconds,
+                max_backoffs_total=max_calls_total,
+                backoff_window_after_first_call_in_seconds=retry_window_after_first_call_in_seconds,
             )
 
-            for retry_action in retry_state:
-                if isinstance(retry_action, DoCall):
-                    try:
-                        return await f(*args, **kwargs)
-
-                    except retry_on_exceptions as e:
-                        last_exception = e
+            while True:
+                try:
+                    return await f(*args, **kwargs)
+                except retry_on_exceptions as e:
+                    last_exception = e
```

Comment:
This could be simplified to: 

```py 

                try:
                    return await f(*args, **kwargs)
                except retry_on_exceptions:
                    sleep_seconds = backoff_calculator.get_backoff()
                    if sleep_seconds is None:
                        raise
```

Though we want to add exception chaining as a follow-up I don't mind if we keep it like this for now.

Reply:
Let's keep as is, refactor is next on my list.